src/mscorlib/shared/System/IO/Path.Windows.cs

```diff
@@ -117,12 +117,17 @@ public static string GetFullPath(string path, string basePath)
                 combinedPath = JoinInternal(basePath, path);
             }
 
-            // Device paths are normalized by definition, so passing something of this format
-            // to GetFullPath() won't do anything by design. Additionally, GetFullPathName() in
-            // Windows doesn't root them properly. As such we need to manually remove segments.
+            int skip = PathInternal.GetRootLength(combinedPath);
+
+            // Device paths are normalized by definition, so passing something of this format (i.e. \\?\C:\.\tmp, \\.\C:\foo)
+            // to Windows APIs won't do anything by design. Additionally, GetFullPathName() in Windows doesn't root
+            // them properly. As such we need to manually remove segments and not use GetFullPath(). We treat "\.." , "\." and "\\" as a
+            // relative segment. In cases like "\\?\C:\.\" and "\\?\C:\..\", we set length of "\\?\C:" (Root Length - 1) as the skip length
+            // for RemoveRelativeSegments. Therefore, the first segment for RemoveRelativeSegments will start with a leading separator.
```

Comment:
This explains what we're trying to do better. It does raise another question, however, and that is why does it matter? Why does the path have to start with a separator after skip?  What would this do for a path of `\\?\C:`  or `\\?\DriveOne` if it could get here?

My guess is that part of the problem here is that we're treating the inputs differently between Unix/Windows. Presumably we should be passing the root length in for Unix as well (e.g. `1`) and doing the logic to _always_ move back one if `-1` is a separator? (`// We want to collapse the first separator past the root presuming the root actually ends in a separator`).

Reply:
What I'm saying here also gets back to what I've suggested before- renaming `skip` to `rootLength` and using `skip` internally in the method would help clarify intent I think.